### ISSUE TITLE
Wrong type of this context for super calls on static functions

### ISSUE THREAD

**Issue by RobinBuschmann** [Microsoft/TypeScript#20264]
**TypeScript Version:**  2.7.0

When calling a static superclass function the `this` context is not recognized correctly. `this` seems to be type of the superclass instead of the calling class type.

**Code**
```ts
class Parent {
  static doWork<T extends Parent>(this: (new () => T)): T {
    return new this();
  }
}

class Child extends Parent { // throws TS2417
  name: string;
  static doWork() {
    return super.doWork();
  }
  static doWorkForChild(): Child {
    return super.doWork(); // throws TS2322
  }
}
```

From my understanding `super.doWork();` should be equivalent to `super.doWork.call(this);`. While `super.doWork.call(this);` does not throw `TS2417`

**Expected behavior:**
Should not throw compiler error.

**Actual behavior:**
Throws:
```
TS2417: Class static side 'typeof Child' incorrectly extends base class static side 'typeof Parent'.
  Types of property 'doWork' are incompatible.
    Type '() => Parent' is not assignable to type '<T extends Parent>(this: new () => T) => T'.
      Type 'Parent' is not assignable to type 'T'.
```
and
```
TS2322:Type 'Parent' is not assignable to type 'Child'.
  Property 'name' is missing in type 'Parent'.
```

**Comment by mhegazy** [comment]
`this` in a static context is the current class, and  not any derived class. the real issue here is that TS does not force derived classes to have a matching constructor signature as their parent class. and i absence of this constraint, there is no way to guarantee that a generic `new this()` is safe, where `this` refers to a more derived class, since you do not know what constructor arguments that derived class has.

**Comment by RobinBuschmann** [comment]
Thanks for your answer. But can you explain why `this` is always the current class and cannot be the derived one? I see no difference between static and non-static functions here. Since `es2015` static functions are inherited via prototype chaining as well. 
Probably my example was a little bit misleading. What about the following (here the constructors doesn't matter at all):
```ts
class Parent {
  static getMe(): typeof Parent {
    return this;
  }
}

class Child extends Parent {
  static id: number;

  static getMe(): typeof Child {
    return super.getMe(); // TS2322:Type 'typeof Parent' is not assignable to type 'typeof Child'.  Property 'id' is missing in type 'typeof Parent'.
  }
}
```
